File: Perfect_Squares.java
/**
 * Problem: https://leetcode.com/problems/perfect-squares/
 */
public class Solution {
    
    private int[] mark;
    
    public int numSquares(int n) {
        int x = (int)Math.sqrt(n);
        mark = new int[n+1];
        for(int i=0; i<=n; i++) mark[i] = -1;
        for(int i = (int)Math.sqrt(n); i >= 1; i--) {
            mark[i*i] = 1;
        }
        return square(n);
    }
    
    private int square(int left) {
        if(mark[left] != -1) return mark[left];
        mark[left] = Integer.MAX_VALUE;
        for(int x = (int)Math.sqrt(left); x >=1; x--){
            mark[left] = Math.min(mark[left], 1 + square(left - x * x));
        }
        return mark[left];
    }
}

File: Swift/perfect-squares.swift
/**
 * https://leetcode.com/problems/perfect-squares/
 * 
 * 
 */ 
// Date: Sat Jun 27 11:19:12 PDT 2020
class Solution {
    func numSquares(_ n: Int) -> Int {
        var steps = Array(repeating: Int.max, count: n + 1)
        steps[0] = 0
        for x in 1 ... n {
            var minStep = steps[x]
            var dx = 1
            while (x - dx * dx) >= 0 {
                minStep = min(minStep, 1 + steps[x - dx * dx])
                dx += 1
            }
            steps[x] = minStep
        }
        
        return steps[n]
    }
}